import re
import pandas
# from pytools.persistent_dict import PersistentDict
# size estimation from kmers as well with jellyfish?
# storage = PersistentDict("qcquickie_storage")
# alternative is to do a quick qc on basepairs followed by a more rigorous assembly

configfile: os.path.join(os.path.dirname(workflow.snakefile), "../config/config.yaml")
# requires --config R1_reads={read_location},R2_reads={read_location}
sample = config["Sample"]
R1 = config["R1_reads"],
R2 = config["R2_reads"],

# my understanding is all helps specify final output
onsuccess:
    print("Workflow complete")
    output = ["qcquickie_status.txt"]
    with open(output[0], "w") as status:
        status.write("Success\n")
onerror:
    print("Workflow error")
    output = ["qcquickie_status.txt"]
    with open(output[0], "w") as status:
        status.write("Failure\n")

rule run__qcquickie:
    message:
        "Denovo assembly of normalized reads"
    input:
        reads = (R1, R2)
    output:
        "qcquickie_done"
    params:
        options = "".join(config["tadpole"]["options"]),
        adapter = "/tools/git.repositories/SerumQC-private/DB/adapters.fasta"
    log:
        "qcquickie/qcquickie.log"
    shell:
        """
        mkdir qcquickie
        cd qcquickie
        bbduk.sh in={input.reads[0]} in2={input.reads[1]} out=filtered.fq ref={params.adapter} ktrim=r k=23 mink=11 hdist=1 tbo minavgquality=14 &>> {log}
        kraken -db /srv/data/DB/kraken/minikraken_20171019_8GB/ --fastq-input filtered.fq | kraken-report -db /srv/data/DB/kraken/minikraken_20171019_8GB/ 1> mini.report 2> {log}
        python /tools/git.repositories/Bracken/est_abundance.py -i mini.report -k /srv/data/DB/kraken/minikraken_20171019_8GB/minikraken_8GB_100mers_distrib.txt -o bracken.txt &>> kraken.log
        bbmerge.sh in=filtered.fq out=merged.fq outu=unmerged.fq &>> {log}
        tadpole.sh in=merged.fq,unmerged.fq out=contigs.fa &>> qcquickie.log
        bbwrap.sh ref=contigs.fa in=merged.fq,unmerged.fq out=contigs.sam append &>> {log}
        pileup.sh in=contigs.sam out=contigs_min.pileup covwindow=1 covwindowavg=10 overwrite=true &>> {log}
        pileup.sh in=contigs.sam out=contigs_rec.pileup covwindow=1 covwindowavg=25 overwrite=true &>> {log}
        filterbycoverage.sh in=contigs.fa cov=contigs_min.pileup minc=10 &>> {log}
        filterbycoverage.sh in=contigs.fa cov=contigs_rec.pileup minc=25 &>> {log}
        sketch.sh in=contigs.fa out=contigs.sketch &>> {log}
        cd ..
        touch qcquickie_done
        """
